man/asmline.1
.TH ASMLINE 1 2022-01-21 GNU

.SH NAME
asmline \- assemblyline tool 

.SH SYNOPSIS
.B asmline
[OPTIONS]...
path/to/file.asm

.SH DESCRIPTION
.B asmline 
Generates machine code from a file or stdin containing x86_64 assembly instructions. Machine code could be directly executed without the need for an executable file format. 

.SH OPTIONS

.TP
.BR \-r[=LEN] ", " \-\-return[=LEN]
Assembles given code. Then executes it with six parameters to heap-allocated memory. Each pointer points to an array of LEN 64-bit elements which can be dereferenced in the asm-code, where LEN defaults to 10. After execution, it prints out the contents of the return (rax) register and frees the heap-memory.

.TP
.BR \-\-rand
Implies -r and will additionally initialize the memory from with random data. 
.br
-r=11 can be used to alter LEN.

.TP
.BR \-p ", " \-\-print
The corresponding machine code will be printed to stdout in hex form. Output is similar to `objdump`: Byte-wise delimited by space and linebreaks after 7 bytes. If-c is given, the chunks are delimited by '|' with each chunk on one line.

.TP
.BR \-P ", " \-\-printfile " " \fIFILENAME
The corresponding machine code will be printed to \fIFILENAME\fR in binary form. Can be set to '/dev/stdout' to write to stdout.

.TP
.BR \-c ", " \-\-chunk " " \fICHUNK_SIZE>1
Sets a given \fICHUNK_SIZE\fR boundary in bytes. Nop padding will be used to ensure no instruction opcode will cross the specified \fICHUNK_SIZE\fR boundary.

.TP
.BR \-b ", " \-\-breaks " " \fICHUNK_BOUNDARY>1

Given a \fICHUNK_BOUNDARY\fR asmline will count the number of instructions where their opcode crosses the specified \fICHUNK_BOUNDARY\fR size in bytes.

.TP
.BR \-o ", " \-\-object " " \fIFILENAME
Generates a binary file from path/to/file.asm called \fIFILENAME\fR.bin in the current directory.

.TP
.BR \-\-nasm\-mov\-imm
Enables nasm-style mov-immediate register-size handling. where if immediate size for mov is less than or equal to max signed 32-bit assemblyline  will emit code to mov to 32-bit register rather than 64-bit.
.br
\fBThat is:\fR "mov rax, 0x7fffffff" as "mov eax, 0x7fffffff" -> b8 ff ff ff 7f
.br
\fBNOTE:\fR rax got optimized to eax for faster immediate to register transfer and produces a shorter instruction.

.TP
.BR \-\-strict\-mov\-imm
Enables strict mov-imm thereby disabling nasm-style mov-immediate register-size handling.
.br
\fBex:\fR even if immediate size for mov is less than or equal to max signed 32-bit
.br
    assemblyline will pad the immediate to fit 64bit.
.br
\fBThat is:\fR "mov rax,0x7fffffff" as "mov rax,0x000000007fffffff" 
.br
          -> 48 b8 ff ff ff 7f 00 00 00 00

.TP
.BR \-\-smart\-mov\-imm
Assemblyline will check the immediate value for leading 0's and thus allows manual optimizations. Immediate must be zero padded to 64-bits exactly to ensure it will not optimize. This is currently set as default.
.br
\fBex:\fR "mov rax, 0x000000007fffffff" -> 48 b8 ff ff ff 7f 00 00 00 00
.br
    "mov rax, 0x7fffffff" -> b8 ff ff ff 7f


.TP
.BR \-\-nasm\-sib\-index\-base\-swap
In SIB addressing if the index register is esp or rsp then the base and index registers will be swapped as this is because the stack pointer register is not scalable in SIB. This is currently set as default.
.br
\fBThat is:\fR "lea r15, [rax+rsp]" interpreted as "lea r15, [rsp+rax]"
.br
         -> 4c 8d 3c 04
.br
         "lea r15, [rsp+rsp]" will produce an error because 
.br
         base and index cannot be swapped)

.TP
.BR \-\-strict\-sib\-index\-base\-swap
In SIB addressing the base and index registers will not be swapped even if the index register is esp or rsp.
.br
\fBThat is:\fR "lea r15, [rax+rsp]" will be interpreted as "lea r15, [rax+riz]"
.br
         -> 4c 8d 3c 20
.br
\fBNOTE:\fR riz is a pseudo-register evaluated by GCC to constant 0 and therefore cannot be used in assemblyline as a string ie. assembling "lea r15, [rax+riz]" is invalid "lea r15, [rsp+rsp]" will produce an error (invalid instruction)


.TP
.BR \-\-nasm\-sib\-no\-base
In SIB addressing if there is no base register present and scale is equal to 2; the base register will be set to the index register and the scale will be reduced to 1.
.br
\fBThat is:\fR "lea r15, [2*rax]" -> "lea r15, [rax+1*rax]"
.br
         -> 4c 8d 3c 00

.TP
.BR \-\-strict\-sib\-no\-base
In SIB addressing when there is no base register present the index and scale would not change regardless of scale value.
.br
\fBThat is:\fR "lea r15, [2*rax]" -> "lea r15, [2*rax]"
.br
         -> 4c 8d 3c 45 00 00 00 00

.TP
.BR \-\-nasm\-sib
Equivalent to \fB--nasm-sib-no-base\fR \fB--nasm-sib-index-base-swap\fR.
.br

.TP
.BR \-\-strict\-sib
Equivalent to \fB--strict-sib-no-base\fR \fB--strict-sib-index-base-swap\fR.
.br

.TP
.BR \-n ", " \-\-nasm
Equivalent to \fB--nasm-mov-imm\fR \fB--nasm-sib-index-base-swap\fR \fB--nasm-sib-no-base\fR.
.br

.TP
.BR \-t ", " \-\-strict
Equivalent to \fB--strict-mov-imm\fR \fB--strict-sib-index-base-swap\fR \fB--strict-sib-index-base-swap\fR.

.TP
.BR \-s ", " \-\-smart
Equivalent to \fB--smart-mov-imm\fR.
.br 

.TP
.BR \-h ", " \-\-help
Prints usage information to stdout and exits.
.TP
.BR \-v ", " \-\-version
Prints version information to stdout and exits.

.SH SEE ALSO
.B libassemblyline(3)

Changelog
version 1.2.2-release (2022-03-29)

	- added support for neg: Two's Complement Negation

	- fixed a bug where shift and rotate instruction produce the wrong opcode 
	  when immediate is zero 

	- fixed a bug when using sib addressing with no base and a single operand

	- asmline can now count the number of instructions that breaks a chunk boundary

	- the number of instructions that break a chunk boundary can be counted when 
	  assembling a file. 

	- adding `cpuid`-instruction
	  


version 1.2.1-release (2022-02-23)
	
	- Fixed bugs regarding -imm/8


version 1.2.0-release (2022-02-19)
	
	- binary file can now be generated from libassemblyline 
	  with asm_create_bin_file()

	- reformatted man page for wide screen

	- added support for all operand encodings for and, or, xor, sub, and add
	  instructions

	- added support for all setcc instructions - set byte on condition

	- added a command-line option in asmline for setting 
	  assembly mode to NASM '$ asmline --nasm-sib-no-base' where:
	  in SIB addressing if there is no base register present and scale is equal to 2; 
	  the base register will be set to the index register and the scale will be reduced to 1.
	  That is: "lea r15, [2*rax]" -> "lea r15, [rax+1*rax]" -> 4c 8d 3c 00

	- added a command-line option in asmline for setting 
	  assembly mode to STRICT '$ asmline --strict-sib-no-base' where:
	  in SIB addressing when there is no base register present the index and scale would
	  not change regardless of scale value.  
	  That is: "lea r15, [2*rax]" -> "lea r15, [2*rax]" -> 4c 8d 3c 45 00 00 00 00


version 1.1.0-release (2022-01-21)

	- added support for scaled index addressing mode
	  with syntax [base + scale*index + offset] or
	  [base + index*scale + offset]
	  ex: "add rax, [rcx+4*rbp+0x8]" or "add rax, [rcx+rbp*4+0x8]"

	- added support for MMX, SSE2, AVX, and AVX2 instruction sets

	- added support for pointers: byte, word, dword, and qword  

	- added support for `mov [MEM], IMM` such as `mov [rsi], "0xADE1A1DE"`
	
	- added different modes of register size handling for 
	  "mov register, immediate" instruction: NASM, STRICT, SMART

	- improved 32-bit register supportability:
	  using eax-r15b for memory addressing in vector extension now 
	  produces the correct prefix (67H)
	
	- improved intel x86_64 instruction supportability:
	  source code is now more inline with the intel manual, therefore
	  more instructions can be supported in the near future

	- added a command-line option in asmline for setting 
	  assembly mode to NASM '$ asmline -n' where:
	  if immediate size for mov is less than or equal to max signed 32-bit
	  asmline will emit code to mov to 32-bit register rather than 64-bit.
	  That is: "mov rax, 0x7fffffff" as "mov eax, 0x7fffffff" -> b8 ff ff ff 7f
	  note: rax got optimized to eax for faster immediate to register transfer and
	  produces a shorter instruction. 

	- added a command-line option in asmline for setting 
	  assembly mode to STRICT '$ asmline -t' where:  
	  ex: even if immediate size for mov is less than or equal to max signed 32-bit
	  pad the immediate to fit 64bit.
	  That is: "mov rax,0x7fffffff" as "mov rax,0x000000007fffffff" 
	    -> 48 b8 ff ff ff 7f 00 00 00 00

	- added a command-line option in asmline for setting  
	  assembly mode to SMART '$ asmline -s' where:  
	  asmline will check the immediate value for leading 0's and thus allows manual
	  optimizations. Immediate must be zero padded to 64-bits exactly to ensure 
	  it will not optimize. This is currently set as default.
	  ex: "mov rax, 0x000000007fffffff" ->  48 b8 ff ff ff 7f 00 00 00 00
	      "mov rax, 0x7fffffff"         ->  b8 ff ff ff 7f

	- assemblyline functions without asm_ prefix are now deprecated

	- simplified the process for adding new instructions
	
	- improvements to autotools, including use of pkg-config. That enables
	  other autotools-based projects to include assemblyline as such:
	  add to their `configure.ac`:
	  ```
		AC_SUBST([LIBASSEMBLYLINE_CFLAGS])
		AC_SUBST([LIBASSEMBLYLINE_LIBS])
		PKG_CHECK_MODULES([LIBASSEMBLYLINE], [assemblyline >= 1.1.0])
	  ```

	- tools/asmline dynamically links to libassemblyline.so

	- tools/asmline's -r[=LEN] option now runs with six arguments, which point
	  to heap-memory. Each argumenty points to LEN (defaults to ten) 64-bit
	  elements. It also features now --rand, which initializes those pointers to
	  random data.

	- added `asmline 1` man-page

	- added man-pages for all the functions from libassemblyline (same content
	  as libassemblyline)
	

version 1.0.5-release (2021-09-29)

	- FIX another bug where 'mov reg, imm' results in a wrong instruction


version 1.0.4-release (2021-09-27)
	
	- FIX bug where 'mov reg, imm' is interpreted as  movntqda
	
	- updated documentation
	
	- implemented register size optimization for mov instruction 
	  to match nasm assembly style
	
	- added man page
	
	- added -v -h options to asmline


version 1.0.3-release (2021-09-16)

	- FIX rdtsc

	- ADD rdtscp,
	
	- ADD ror (ri encoding only for now)
	
	- code maintenance
	
	- added library versioning


version 1.0.2-release (2021-09-09)

	- added support for the prefetch instruction
	
	- added support for all registers (prefetch)


version 1.0.1-release (2021-09-09)

	- added support for the rorx instruction
	
	- updated documentation


version 1.0.0-release (2021-08-26)
	
	- initial release

© 2022 GitHub, Inc.


tools/asmline.c
/**
 * Copyright 2022 University of Adelaide
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*command-tool for generating machine code from a x64 assembly file*/
#include <assemblyline.h>
#if HAVE_CONFIG_H
#include <config.h> // from autotools
#endif
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define DEFAULT_ARG_LEN 10
#define BUFFER_SIZE 100
enum OUTPUT { NONE, BIN_FILE, GENERIC_FILE };
enum run { DONT_RUN = 0, RUN = 1, RUN_RAND = 2 };

typedef enum {

  NASM_MOV_IMM = 128,
  STRICT_MOV_IMM,
  NASM_SIB,
  STRICT_SIB,
  NASM_SIB_INDEX_BASE_SWAP,
  STRICT_SIB_INDEX_BASE_SWAP,
  NASM_SIB_NO_BASE,
  STRICT_SIB_NO_BASE,
  SMART_MOV_IMM
} asm_options;

const char *asm_version = PACKAGE_STRING;
static int mov_imm = 0;
static int sib_all = 0;
static int sib_swap = 0;
static int sib_no_base = 0;

void err_print_usage(char *error_msg) {
  fprintf(
      stderr,
      "%s\nUsage: asmline "
      "[OPTIONS]... path/to/file.asm\n\n"

      "  -r[=LEN], --return[=LEN]\n"
      "\tAssembles given code. Then executes it with six parameters to "
      "heap-allocated memory.\n\tEach pointer points to an array of LEN 64-bit "
      "elements which can be dereferenced \n\tin the asm-code, where LEN "
      "defaults "
      "to 10.\n\tAfter execution, it prints out the contents of the return "
      "(rax) register and frees  \n\tthe heap-memory.\n\n"

      "  --rand \n"
      "\tImplies -r and will additionally initialize the memory from with "
      "random data. \n\t-r=11 can be used to alter LEN.\n\n"

      "  -p, --print\n"
      "\tThe corresponding machine code will be printed to stdout in hex "
      "form.\n"
      "\tOutput is similar to `objdump`: Byte-wise delimited by space and "
      "linebreaks after 7 bytes.\n\tIf -c is given, the chunks are "
      "delimited by '|' and each chunk is on one line.\n\n"

      "  -P, --printfile FILENAME\n"
      "\tThe corresponding machine code will be printed to FILENAME in binary "
      "form.\n\tCan be set to '/dev/stdout' to write to stdout.\n\n"

      "  -o, --object FILENAME\n"
      "\tThe corresponding machine code will be printed to FILENAME.bin in "
      "binary.\n\n"

      "  -c, --chunk CHUNK_SIZE>1\n"
      "\tSets a given CHUNK_SIZE boundary in bytes. Nop padding will be used "
      "to ensure no instruction\n"
      "\topcode will cross the specified CHUNK_SIZE boundary.\n\n"

      "  -b, --breaks CHUNK_BOUNDARY>1\n"
      "\tGiven a CHUNK_BOUNDARY, counts the number of instructions where\n"
      "\ttheir opcode crosses the specified CHUNK_BOUNDARY size in bytes.\n\n"

      "  --nasm-mov-imm\n"
      "\tEnables nasm-style mov-immediate register-size handling.\n"
      "\tex: if immediate size for mov is less than or equal to max "
      "signed 32 bit assemblyline\n"
      "\t    will emit code to mov to the 32-bit register rather than 64-bit.\n"
      "\tThat is: \"mov rax,0x7fffffff\" as \"mov eax,0x7fffffff\" "
      "-> b8 ff ff ff 7f\n"
      "\tnote: rax got optimized to eax for faster immediate to register "
      "transfer\n"
      "\t      and produces a shorter instruction\n\n"

      "  --strict-mov-imm\n"
      "\tDisables nasm-style mov-immediate register-size handling.\n"
      "\tex: even if immediate size for mov is less than or equal to max "
      "signed 32 bit assemblyline.\n"
      "\t    will pad the immediate to fit 64-bit\n"
      "\tThat is: \"mov rax,0x7fffffff\" as \"mov rax,0x000000007fffffff\"\n"
      "\t          -> 48 b8 ff ff ff 7f 00 00 00 00\n\n"

      "  --smart-mov-imm\n"
      "\tThe immediate value will be checked for leading 0's.\n"
      "\tImmediate must be zero padded to 64-bits exactly to ensure\n"
      "\tit will not optimize. This is currently set as default.\n"
      "\tex: \"mov rax, 0x000000007fffffff\" ->  48 b8 ff ff ff 7f 00 00 00 "
      "00\n\n"

      "  --nasm-sib-index-base-swap\n"
      "\tIn SIB addressing if the index register is esp or rsp then\n"
      "\tthe base and index registers will be swapped.\n"
      "\tThat is: \"lea r15, [rax+rsp]\" -> \"lea r15, [rsp+rax]\"\n\n"
      "  --strict-sib-index-base-swap\n"
      "\tIn SIB addressing the base and index registers will not be swapped\n"
      "\teven if the index register is esp or rsp.\n\n"

      "  --nasm-sib-no-base\n"
      "\tIn SIB addressing if there is no base register present and scale\n"
      "\tis equal to 2; the base register will be set to the index register\n"
      "\tand the scale will be reduced to 1.\n"
      "\tThat is: \"lea r15, [2*rax]\" -> \"lea r15, [rax+1*rax]\"\n\n"
      "  --strict-sib-no-base\n"
      "\tIn SIB addressing when there is no base register present the index\n"
      "\tand scale would not change regardless of scale value.\n"
      "\tThat is: \"lea r15, [2*rax]\" -> \"lea r15, [2*rax]\"\n\n"

      "  --nasm-sib\n"
      "\tequivalent to --nasm-sib-index-base-swap --nasm-sib-no-base\n\n"
      "  --strict-sib\n"
      "\tequivalent to --strict-sib-index-base-swap --strict-sib-no-base\n\n"

      "  -n, --nasm\n"
      "\tequivalent to --nasm-mov-imm --nasm-sib\n\n"

      "  -t, --strict\n"
      "\tequivalent to --strict-mov-imm --strict-sib\n\n"

      "  -h, --help\n"
      "\tPrints usage information to stdout and exits.\n\n"

      "  -v, --version\n"
      "\tPrints version information to stdout and exits.\n\n",
      error_msg);
  exit(EXIT_FAILURE);
}

void print_version() {

  printf("%s\n"
         "Copyright 2022 University of Adelaide\n\n"
         "Licensed under the Apache License, Version 2.0 (the \"License\");\n"
         "You may obtain a copy of the License at\n\n"
         "\thttp://www.apache.org/licenses/LICENSE-2.0\n\n"
         "Unless required by applicable law or agreed to in writing, software\n"
         "distributed under the License is distributed on an \"AS IS\" BASIS,\n"
         "WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or "
         "implied.\n"
         "See the License for the specific language governing permissions and\n"
         "limitations under the License.\n\n"
         "Written by David Wu and Joel Kuepper\n",
         asm_version);
  exit(EXIT_SUCCESS);
}

int check_digit(char *optarg) {
  int len = strlen(optarg);
  for (int i = 0; i < len; i++)
    if (optarg[i] < '0' || optarg[i] > '9')
      return EXIT_FAILURE;
  return EXIT_SUCCESS;
}

void print_chunk_brks(int total_chunk_brks, bool debug, int chunk_boundary) {

  printf("%d", total_chunk_brks);
  if (debug)
    printf(" instructions break a chunk boundary of %d bytes", chunk_boundary);
  printf("\n");
}

void execute_get_ret_value(void *function, int arglen, enum run mode) {
  uint64_t *arguments[6];

  uint64_t result = 0;
  if (arglen == 0) {
    uint64_t (*f)() = function;
    result = f();
  } else {
    // allocate 6 args with arglen uint64_t's
    for (int arg_idx = 0; arg_idx < 6; arg_idx++) {
      arguments[arg_idx] = calloc(arglen, sizeof(uint64_t));
      if (mode & RUN_RAND)
        for (int qword_idx = 0; qword_idx < arglen; qword_idx++) {

          uint64_t rand_val = rand();           // lo_limb
          rand_val |= ((uint64_t)rand()) << 32; // hi_limb
          arguments[arg_idx][qword_idx] = rand_val;
        }
    }
    // cast
    uint64_t (*f)(uint64_t *, uint64_t *, uint64_t *, uint64_t *, uint64_t *,
                  uint64_t *) = function;
    // call
    result = f(arguments[0], arguments[1], arguments[2], arguments[3],
               arguments[4], arguments[5]);
    // free args
    for (int i = 0; i < 6; i++)
      free(arguments[i]);
  }
  printf("\nthe value is 0x%lx\n", result);
}

int main(int argc, char *argv[]) {

  int opt;
  enum OUTPUT create_bin = NONE;
  char bin_ext[] = ".bin";
  char *param_file = NULL;
  char *bin_file = NULL;
  char *write_file = NULL;
  int chunk_boundary = 0;
  int total_chunk_brks = 0;
  bool debug = false;

  // for running (with arguments)
  enum run get_ret = DONT_RUN;
  int arglen = DEFAULT_ARG_LEN;

  struct option long_options[] = {
      /* These options set a flag. */
      {"nasm-mov-imm", no_argument, &mov_imm, NASM_MOV_IMM},
      {"strict-mov-imm", no_argument, &mov_imm, STRICT_MOV_IMM},
      {"smart-mov-imm", no_argument, &mov_imm, SMART_MOV_IMM},
      {"nasm-sib", no_argument, &sib_all, NASM_SIB},
      {"strict-sib", no_argument, &sib_all, STRICT_SIB},
      {"nasm-sib-index-base-swap", no_argument, &sib_swap,
       NASM_SIB_INDEX_BASE_SWAP},
      {"strict-sib-index-base-swap", no_argument, &sib_swap,
       STRICT_SIB_INDEX_BASE_SWAP},
      {"nasm-sib-no-base", no_argument, &sib_no_base, NASM_SIB_NO_BASE},
      {"strict-sib-no-base", no_argument, &sib_no_base, STRICT_SIB_NO_BASE},
      {"version", no_argument, 0, 'v'},
      {"help", no_argument, 0, 'h'},
      {"rand", no_argument, (int *)&get_ret, RUN_RAND},
      {"return", optional_argument, 0, 'r'},
      {"print", no_argument, 0, 'p'},
      {"printfile", required_argument, 0, 'P'},
      {"nasm", no_argument, 0, 'n'},
      {"strict", no_argument, 0, 't'},
      {"smart", no_argument, 0, 's'},
      {"chunk", required_argument, 0, 'c'},
      {"breaks", required_argument, 0, 'b'},
      {"object", required_argument, 0, 'o'},
      {0, 0, 0, 0}};

  /* getopt_long stores the option index here. */
  int option_index = 0;

  if (argc < 2 && isatty(fileno(stdin)))
    err_print_usage("Error: invalid number of arguments\n");

  assemblyline_t al = asm_create_instance(NULL, 0);
  while ((opt = getopt_long(argc, argv, "hvr::ntspP:c:b:o:", long_options,
                            &option_index)) != -1) {
    switch (opt) {
    case 0:
      // intentionally blank for the options with implicit flag setting by
      // getopt_long, like --rand
      break;
    case 'v':
      print_version();
      break;
    case 'h':
      err_print_usage("");
      break;
    case 'r':
      // if there is a optional argument, try to parse it and set the arg len
      if (optarg && !check_digit(optarg))
        arglen = atoi(optarg);
      get_ret |= RUN;
      break;
    case 'p':
      debug = true;
      asm_set_debug(al, true);
      break;
    case 'n':
      asm_set_all(al, NASM);
      break;
    case 't':
      asm_set_all(al, STRICT);
      break;
    case 's':
      asm_set_all(al, SMART);
      break;
    case 'c':
      if (check_digit(optarg))
        err_print_usage("Error: [-c CHUNK_SIZE>1] expects an integer\n");
      int chunk_size = atoi(optarg);
      asm_set_chunk_size(al, chunk_size);
      break;

    case 'b':
      if (check_digit(optarg))
        err_print_usage("Error: [-b CHUNK_BOUNDARY>1] expects an integer\n");
      chunk_boundary = atoi(optarg);
      break;

    case 'P':
      create_bin = GENERIC_FILE;
      param_file = optarg;
      break;
    case 'o':
      if (strchr(optarg, '.'))
        err_print_usage("elf filename cannot have an extension\n");
      create_bin = BIN_FILE;
      param_file = optarg;
      break;

    default: /* '?' */
      if (mov_imm || sib_swap || sib_no_base || sib_all)
        break;
      err_print_usage("");
    }
  }

  switch (mov_imm) {
  case 0:
    break;
  case NASM_MOV_IMM:
    asm_mov_imm(al, NASM);
    break;
  case STRICT_MOV_IMM:
    asm_mov_imm(al, STRICT);
    break;
  case SMART_MOV_IMM:
    asm_mov_imm(al, SMART);
    break;
  default:
    break;
  }

  switch (sib_swap) {
  case 0:
    break;
  case NASM_SIB_INDEX_BASE_SWAP:
    asm_sib_index_base_swap(al, NASM);
    break;
  case STRICT_SIB_INDEX_BASE_SWAP:
    asm_sib_index_base_swap(al, STRICT);
    break;
  default:
    break;
  }

  switch (sib_no_base) {
  case 0:
    break;
  case NASM_SIB_NO_BASE:
    asm_sib_no_base(al, NASM);
    break;
  case STRICT_SIB_NO_BASE:
    asm_sib_no_base(al, STRICT);
    break;
  default:
    break;
  }

  switch (sib_all) {
  case 0:
    break;
  case NASM_SIB:
    asm_sib_no_base(al, NASM);
    asm_sib_index_base_swap(al, NASM);
    break;
  case STRICT_SIB:
    asm_sib_no_base(al, STRICT);
    asm_sib_index_base_swap(al, STRICT);
    break;
  default:
    break;
  }

  if (optind >= argc) {
    // check is stdin is provided via pipe
    if (!isatty(fileno(stdin))) {
      char *line = NULL;
      size_t size = BUFFER_SIZE;
      if (!chunk_boundary) {
        while (getline(&line, &size, stdin) != -1) {
          if (asm_assemble_str(al, line)) {
            fprintf(stderr, "failed to assemble instruction: %s\n", line);
            exit(EXIT_FAILURE);
          }
        }
      } else {
        while (getline(&line, &size, stdin) != -1) {
          int chunk_brks = 0;
          if (asm_assemble_string_counting_chunks(al, line, chunk_boundary,
                                                  &chunk_brks)) {
            fprintf(stderr, "failed to assemble instruction: %s\n", line);
            exit(EXIT_FAILURE);
          }
          total_chunk_brks += chunk_brks;
        }
        print_chunk_brks(total_chunk_brks, debug, chunk_boundary);
      }
      free(line);
    } else
      err_print_usage("Error: Expected path/to/file.asm after options\n");
  } else {
    if (!chunk_boundary) {
      if (asm_assemble_file(al, argv[optind])) {
        fprintf(stderr, "failed to assemble file: %s\n", argv[optind]);
        exit(EXIT_FAILURE);
      }
    } else {
      if (asm_assemble_file_counting_chunks(al, argv[optind], chunk_boundary,
                                            &total_chunk_brks)) {
        fprintf(stderr, "failed to assemble file: %s\n", argv[optind]);
        exit(EXIT_FAILURE);
      }
      print_chunk_brks(total_chunk_brks, debug, chunk_boundary);
    }
  }

  if (get_ret != DONT_RUN) {
    // initialize the randomiser if needed
    if (get_ret & RUN_RAND)
      srand(time(NULL));
    void *func = asm_get_code(al);
    execute_get_ret_value(func, arglen, get_ret);
  }

  switch (create_bin) {
  case NONE:
    exit(EXIT_SUCCESS);
    break;
  case BIN_FILE: {
    size_t bin_file_len = strlen(param_file) + strlen(bin_ext) + 1;
    bin_file = calloc(bin_file_len, sizeof(char));
    sprintf(bin_file, "%s%s", param_file, bin_ext);
    write_file = bin_file;
  } break;
  case GENERIC_FILE:
    write_file = param_file;
    break;
  }
  if (asm_create_bin_file(al, write_file)) {
    fprintf(stderr, "failed to create %s\n", param_file);
    exit(EXIT_FAILURE);
  }
  exit(EXIT_SUCCESS);
}